[WIP] Enable Vault UI

arianvp opened the pull request:
###### Motivation for this change
Since version 0.10, Vault ships with an administrator UI. This PR allows us to use that.
This was a bit of task to get working!

The new consul version also packages its UI this way. So when we want to bump consul from the current (pretty old) version, we can use the same groundwork we did here. So that's good news.

Including the assets has barely any impact on the closure size:
```[arian@ryzen:~/Projects/nixpkgs]$ nix path-info -S $(nix-build -A vault)
these paths will be fetched (13.92 MiB download, 76.72 MiB unpacked):
  /nix/store/5ii0rjvfj6s9260g1mwyrx8bmyqildh9-vault-0.11.2
copying path '/nix/store/5ii0rjvfj6s9260g1mwyrx8bmyqildh9-vault-0.11.2' from 'https://cache.nixos.org'...
/nix/store/5ii0rjvfj6s9260g1mwyrx8bmyqildh9-vault-0.11.2	   83119048


[arian@ryzen:~/Projects/nixpkgs]$ nix path-info -S $(nix-build -A vault)
/nix/store/dwhii30w2xlbs81mpca2kzx8y3byrmrq-vault-0.11.2	   83779736

```



###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

arianvp committed:
vault: Enable vault UI

Enables the vault UI, which ships with vault since version 0.10

arianvp committed:
nixos/vault: Add option to enable the vault web ui

Also adds a test that this option actually works

arianvp committed:
vault: Add arianvp to the maintainers list

arianvp commented:
I'm not sure why the build is failing. It gives a "path is not valid" but I can't reproduce that locally

arianvp commented:
The problem is that consul and vault use yarn.lock instead of  package-lock.json.  Which npm2nix isn't compatible with ... But I could try. Maybe I don't need the exact upstream versions for them.

arianvp commented:
Another problem is that some of the dependencies of Vault's UI are not on `npm` but are on `github` instead.  But I could see if I can try to split that up someway.

arianvp commented:
I disabled import from derivation by explicitly passing a lockfile I
thought.

On Fri, Oct 26, 2018, 02:24 Daiderd Jordan <notifications@github.com> wrote:

> I think the evaluation error is cause by the way yarn2nix works, it uses
> import from derivation which isn't allowed on ofborg or hydra.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/49082#issuecomment-433247751>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAmWo1FoG7q2Sqzn-ujhxx8KuJ3eu4FQks5uolYrgaJpZM4X5Sql>
> .
>

LnL7 commented:
I thought so too, but maybe not? The error definitely looks like an error caused by `restrict-eval` (see man nix.conf)

arianvp commented:
Is patterns like:
https://github.com/NixOS/nixpkgs/pull/49082/files#diff-9f45399180b286c35746a0fcafa076cfR25

disallowed in restricted eval mode?   I.e.   the source of one package is dependent on the source of another package?

arianvp commented:
Ahaaa... It's `importJSON`'ing  `package.json`....  I see.

arianvp commented:
I filed an issue upstream with `yarn2nix` https://github.com/moretea/yarn2nix/issues/83

arianvp committed:
yarn2nix: Remove lib.importJSON from yarn2nix

This method isn't allowed on Hydra. It was only used to
extract metadata from package.json

We remove it, so that the build is pure again.
We should probably upstream this to the yarn2nix repo
as well

arianvp committed:
vault: Use new yarn2nix

arianvp commented:
Apparently `consul` simply vendors the UI assets in their Repo: https://github.com/NixOS/nixpkgs/pull/48714#issuecomment-433454834 . This makes building `consul` a breeze.

I've asked upstream if they would consider doing something similar to `vault`. This means we wouldn't have to vendor 12k lines of yarn dependencies.   https://github.com/hashicorp/vault/issues/5622

arianvp commented:
Unfortunately, they will not fix this upstream. So we'll have to go with `yarn2nix`

arianvp commented:
@Mic92 are there any technical downsides of including this `yarn.nix` file that I'm not overseeing, like for example evaluation time or other issues?  If not, I would prefer to merge this as is, as putting all these dependencies in `node-modules` seems non-trivial to me at this moment.  For the reasons put above in this thread.

Mic92 commented:
This pull request adds 500 kb generated data to our repository. This will stick in our history and requires additional storage for each channel user. I can tell from experience that also git does not like large generated text files. Also we might be able to fix nix evaluation performance it would be impossible to fix our history.
To still make progress on this pull request I suggest the following approach:

1. Fork vault on github
2. Add travis configuration that builds the ui assets and upload them on release tags. 
3. Change the nix build to just take the compiled asset archive instead.

Here is an example project for uploading release assets: https://github.com/Mic92/cntr/blob/master/.travis.yml#L39

arianvp commented:
I lost interest in getting this merged, as I have no interest in maintaining my own fork of vault with my own CI, given it's quite a security-senstive product.
